File: Ruby/test/test_sum_range.rb
require_relative "test_helper"
require_relative "../solution/sum_range"

test_cases = [
  [[["NumArray", "sumRange", "sumRange", "sumRange"], [[[-2, 0, 3, -5, 2, -1]], [0, 2], [2, 5], [0, 5]]], [0, 1, -1, -3]],
]

test_cases.each do |testcase|
  input0, expected = testcase
  commands, params = input0
  num_array = nil
  commands.each_with_index.map do |command, i|
    if command == "NumArray"
      num_array = NumArray.new(params[i][0])
    elsif command == "sumRange"
      output = num_array.sum_range(params[i][0], params[i][1])
      assert_equal(expected[i], output)
    else
      raise "Unknown command #{command}"
    end
  end
end

test_summary

File: Rust/tests/test_sum_range.rs
use leetcode_rust::solution::sum_range::NumArray;

#[test]
fn test_example_1() {
    let input0 = ["NumArray", "sumRange", "sumRange", "sumRange"];
    let input1 = [vec![-2, 0, 3, -5, 2, -1],
        vec![0, 2],
        vec![2, 5],
        vec![0, 5]];
    let expected = [0, 1, -1, -3];

    let mut sln = NumArray::new(input1[0].clone());

    for i in 1..input0.len() {
        let op = &input0[i];
        let params = &input1[i];
        if *op == "NumArray" {
            sln = NumArray::new(params.clone());
        } else if *op == "sumRange" {
            let output: i32 = sln.sum_range(params[0], params[1]);
            assert_eq!(output, expected[i]);
        } else {
            panic!("Unknown operation");
        }
    }
}
